### PR TITLE
Multiband augmentation

### PR DESCRIPTION
**Pull request by Shadiouss**
Dear Sebastian,

I wanted to inform you about the changes I made in the recent pull request. The main updates are as follows:

1. **Transform Function Update:**
   - Replaced the Torchvision method with the Albumentation method for image transformations.
   - Implemented this change in the `utils.py` file.
   - Called the new transformation function in the `train.py` file.

2. **Parameter Settings:**
   - Defined a variable named `aug_pipe` in the `params_and_train.py` file to set the parameters for image augmentation.

3. **Visualization Enhancements:**
   - Added code to the training function to plot the confusion matrix and the classification report at the end of the training loop.

Best,
Shadi

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -3,10 +3,14 @@
 import warnings
 import matplotlib.pyplot as plt
 import numpy as np
+from PIL import Image
```
Comment (benjaminstoeckigt) [inline utils.py]: I think PIL.Image is not gonna be used right?

File: train.py [hunk 2]
```diff
@@ -312,6 +326,7 @@ def train_func (data_path, existing_model, model_path, BATCH_SIZE, visualize_dat
 
         if not enable_regression:
```
Comment (benjaminstoeckigt) [inline train.py:330]: Now i know what you mean with including validation in training pipeline. 
I mean it can be also useful for developing and creating the right model but maybe at least make the whole validation process (starting from 327) optional?
Or for developing pipeline we create a complete new validation function seperated from training code.

File: params_and_main.py [hunk 3]
```diff
@@ -88,13 +88,22 @@
...
+            A.RandomBrightnessContrast( # Randomly changes brightness and contrast of the image with a probability of 0.5.
+                brightness_limit=(-0.1,0.1), 
+                contrast_limit=(-0.1, 0.1), 
+                p=0.5
+            ),
+            A.CoarseDropout(p=0.5), # Randomly masks out rectangular regions in the image with a probability of 0.5.
+
+]) # For more Augmentation options: https://github.com/albumentations-team/albumentations/tree/main#i-am-new-to-image-augmentation
+
+Num = 2 # Number of the augmented images minus from the totla number of batch size (default is 2).
```
Comment (benjaminstoeckigt) [inline params_and_main.py]: i think this means that 2 images within a batch gonna be not transformed - is this meant to be? Maybe i got this wrong but i thought we wanted to transform all images. With batchsize 4 there will be 50 % of images transformed.
 
### see albumentation function (utils.py, line  #327):  ##
Leave the second proportion of the batch unchanged
        for img, mask in zip(batch_img[int(self.Num - len(batch_img)):], batch_mask[int(self.Num - len(batch_img)):]):

File: environment/requirements.txt [hunk 4]
```diff
@@ -1,14 +1,13 @@
 geopandas==0.9.0
-numpy==1.21.1
 pandas==1.3.1
-pillow==8.3.1
-scikit-learn==0.24.2
-scipy==1.7.1
+albumentations==0.3
 fastai==2.5.1
+numpy==1.22.4
```
Comment (benjaminstoeckigt) [inline environment/requirements.txt:5]: hopefully numpy 1.22.4 instaed of 1.21.1 wont make any trouble

**Review by benjaminstoeckigt — CHANGES_REQUESTED** [review]

File: utils.py (hunk 1 above)
Comment (Shadiouss) [reply utils.py]: right, I added some lines to plot some examples during the running time, and then I removed them.. 
So yea, I will remove it.

File: train.py (hunk 2 above)
Comment (Shadiouss) [reply train.py:330]: Do you believe it would improve the code to introduce a new boolean variable, such as 'Evaluate', in the 'params_and_main.py' file? This would allow the user to specify if evaluation should be included after the training pipeline by setting 'Evaluate = True'. Additionally, they could specify which data they want to evaluate from the list of valid scenes, such as ['vali'] or ['test'].

File: params_and_main.py [hunk 5]
```diff
@@ -14,7 +14,7 @@
 import pathlib
 import os
 import warnings
-
+import albumentations as A
```
Comment (SebastianLeh) [inline params_and_main.py:17]: For the future I think it would be good to not have to import albumentations here but in the train.py -> this could be useful for the json approach as we then have to only define 'strings' for the augmentations to be passed and the actual objects to be created in a different module (train.py) -> for now I think its ok like this, too, just as a note

File: params_and_main.py (hunk 3 above)
Comment (SebastianLeh) [inline params_and_main.py]: this comment is a bit confusing to me -> would 1 lead to 3 or 1 image being transformed? I would suggest the number should indicate the number of transformed images and this should be reflected in the comment as '#number of images on which transformations are applied per batch' or so

Comment (SebastianLeh) [inline params_and_main.py]: maybe a more descriptive variable names like 'n__transforms' or 'n_transform_imgs'

**Review by SebastianLeh — CHANGES_REQUESTED** [review]
I would suggest to change the 'Num' variable Name and clarify its comment

File: train.py [hunk 6]
```diff
@@ -325,4 +340,36 @@ def train_func (data_path, existing_model, model_path, BATCH_SIZE, visualize_dat
...
+            plt.ylabel("True")
+            confusion_matrix_path = os.path.join(os.path.dirname(model_path), "confusion_matrix.png")
+            plt.savefig(confusion_matrix_path)
+            plt.show()
+            
+            # Generate and print classification report
+            class_report = classification_report(valid_labels_flat, valid_preds_flat, target_names=classes_name, output_dict=True)
+            df_class_report = pd.DataFrame(class_report).transpose()
+            
+            # Plot classification report
```
Comment (SebastianLeh) [inline train.py]: Did you find out whether this classification report is the same as the confusion matrix? Also, does it solve the problem with classes being unidentifiable in case not all classes are predicted? I remember you once said that the implementation you were using has an issue with showing less classes in case not all are predicted and then its not possible to trace back which class labels refer to which class.

File: train.py (hunk 2 above)
Comment (SebastianLeh) [reply train.py:330]: I think this could make sense, the only risk I see is further bloating the params part. If so, I would try to find a descriptive name like 'save_confusion_matrix' or so.

File: utils.py [hunk 7]
```diff
@@ -235,4 +239,97 @@ def check_and_fill(args, target_len):
...
+        
+        # Process each image and mask in the first proportion of the batch
+        for img, mask in zip(batch_img[:int(self.Num - len(batch_img))], batch_mask[:int(self.Num - len(batch_img))]):
+            # Normalize the image
+            img = img / img.max()
+        
+            # Permute the image dimensions from (C, H, W) to (H, W, C) for albumentations
+            img = img.permute(1, 2, 0)  # Now shape is [W, H, C]
+        
+            print('Applying augmentation')
```
Comment (SebastianLeh) [inline utils.py]: does this mean its printing for each batch that is transformed / not transformed (line 332)? If yes, I think this might be to verbose as there are a lot of print statements during training?

File: utils.py [hunk 8]
```diff
@@ -235,4 +239,97 @@ def check_and_fill(args, target_len):
 
 
 
+## Augmentation
+class SegmentationAlbumentationsTransform(ItemTransform):
```
Comment (SebastianLeh) [inline utils.py:243]: Is there a reason why you used a class? The class seems to make sense, however, I am wondering if we are making the UNet Framework more confusing or fragile if we are now also starting to implement classes, as most functionality is implemented in functions so far.

**Review by SebastianLeh — CHANGES_REQUESTED** [review]
My first comment is more general and the second about the print() verbosity refers to a potential change request.

File: utils.py (hunk 7 above)
Comment (Shadiouss) [reply utils.py]: ah right, I edited this line with the line in the train function .... 

       if transforms:
            print(f"Applying Augmentation on ({Num}) images from ({BATCH_SIZE}) images")
            # Use the imported aug_pipe
            transforms = SegmentationAlbumentationsTransform(aug_pipe, Num=Num)

File: train.py (hunk 6 above)
Comment (Shadiouss) [reply train.py]: No, if the number of actual classes matches the number of predicted classes (as it should in an ideal situation), everything will work correctly. However, if the numbers differ, an error message will be generated, indicating that the number of true labels (number) does not equal the number of predicted labels (number).
But yea I will try to edit this process where we can compare between input classes and output to manage that in the next branch :)

File: params_and_main.py (hunk 3 above)
Comment (Shadiouss) [reply params_and_main.py]: yes, good idea, done!

Comment (benjaminstoeckigt) [reply params_and_main.py]: i would even go for percentage. So in relation to the batch_size. what do you think? otherwise it is a bit random when using different batch sizes e.g. 4, 8, 16, ...

File: train.py [hunk 9]
```diff
@@ -258,12 +259,22 @@ def train_unet(class_weights, dls, architecture, epochs, path, lr, encoder_facto
...
 
         if existing_model is not None:
             existing_model = Path(existing_model)
+        if transforms:
+            print(f"Applying Augmentation on ({BATCH_SIZE - Num}) images from ({BATCH_SIZE}) images")
+            # Use the imported aug_pipe
+            transforms = SegmentationAlbumentationsTransform(aug_pipe, n_transform_imgs=n_transform_imgs)
+        else:
+            # Define a default augmentation pipeline
+            aug_pipe = A.Compose([
```
Comment (benjaminstoeckigt) [inline train.py:278]: Is this necessary? Because 'aug_pipe' is not being passed on. I believe 'transformations = False' given from the initial parameters should be enough for the data_block.

File: params_and_main.py (hunk 3 above)
Comment (Shadiouss) [reply params_and_main.py]: good idea Benny, I have already changed this line in the function:

for img, mask in zip(batch_img[int(self.n_transform_imgs - len(batch_img)):], batch_mask[int(self.n_transform_imgs - len(batch_img)):]):

So now the number that the user will define as  "n_transformed" will be the specific number for augmented!

Comment (Shadiouss) [reply params_and_main.py]: I have changed the code to be the number that indicated will be the same number of transformed images!

File: environment/requirements.txt [hunk 10]
```diff
@@ -1,14 +1,13 @@
 geopandas==0.9.0
-numpy==1.21.1
 pandas==1.3.1
```
Comment (SebastianLeh) [inline environment/requirements.txt:2]: test_comment

File: params_and_main.py [hunk 11]
```diff
@@ -88,13 +89,24 @@
 large_file = False # If predicted probabilities should be stretched to int8 to increase storage capacity
```
Comment (SebastianLeh) [inline params_and_main.py:89]: test_two_lines

**Review by SebastianLeh — CHANGES_REQUESTED** [review]
here's my review (...)

File: environment/requirements.txt [hunk 12]
```diff
@@ -1,14 +1,13 @@
...
-scikit-learn==0.24.2
-scipy==1.7.1
+albumentations==0.3
 fastai==2.5.1
+numpy==1.22.4
 Fiona-1.8.19-cp39-cp39-win_amd64.whl
 GDAL-3.2.3-cp39-cp39-win_amd64.whl
 rasterio-1.2.3-cp39-cp39-win_amd64.whl
 slidingwindow==0.0.14
+seaborn==0.13.2
```
Comment (SebastianLeh) [inline environment/requirements.txt:10]: accepting review

**Review by SebastianLeh — APPROVED** [review]

File: environment/requirements.txt [hunk 13]
```diff
@@ -1,14 +1,13 @@
 geopandas==0.9.0
```
Comment (benjaminstoeckigt) [inline environment/requirements.txt:1]: nice

**Review by benjaminstoeckigt — APPROVED** [review]
nice work

**Review by benjaminstoeckigt — CHANGES_REQUESTED** [review]
test

**Review by benjaminstoeckigt — APPROVED** [review]

**Review by benjaminstoeckigt — CHANGES_REQUESTED** [review]
test

**Review by benjaminstoeckigt — APPROVED** [review]
test

File: environment/requirements.txt [hunk 14]
```diff
@@ -1,14 +1,13 @@
...
-numpy==1.21.1
 pandas==1.3.1
-pillow==8.3.1
-scikit-learn==0.24.2
```
Comment (SebastianLeh) [inline environment/requirements.txt:5]: test_deny_pull_requets

**Review by SebastianLeh — CHANGES_REQUESTED** [review]
test change request

**Review by SebastianLeh — APPROVED** [review]
